internal/envoy/auth/oauth2_filter.go

```diff
@@ -0,0 +1,233 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package auth
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	// xds_core_v3 "github.com/cncf/xds/go/xds/core/v3"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_extensions_filter_http_oauth2_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/oauth2/v3"
+	envoy_extensions_transport_sockets_tls_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/kubeshop/kusk-gateway/pkg/options"
+)
+
+func NewFilterHTTPOAuth2(oauth2Options *options.OAuth2, args *parseAuthOptionsArguments) (*anypb.Any, error) {
+	// Example Input: "https://kubeshop-kusk-gateway-oauth2.eu.auth0.com/oauth/token"
+	// Example Output from url.Hostname(): "kubeshop-kusk-gateway-oauth2.eu.auth0.com"
+	// Example Output from url.Port(): ""
+	url, err := url.Parse(oauth2Options.TokenEndpoint)
+	if err != nil {
+		err = fmt.Errorf("auth.NewFilterHTTPOAuth2: could not determine upstreamServiceHost from oauth2.token_endpoint=%q, %w", oauth2Options.TokenEndpoint, err)
+		args.Logger.Error(err, "oauth2.token_endpoint contains an invalid url, failed to parse url")
+		return nil, err
+	}
+
+	// // cluster := "kubeshop-kusk-gateway-oauth2.eu.auth0.com"
+	// cluster := "eu.auth0.com"
+	// sni := "eu.auth0.com"
+	// // sni := "kubeshop-kusk-gateway-oauth2.eu.auth0.com"
+	// // upstreamServiceHost := "kubeshop-kusk-gateway-oauth2.eu.auth0.com"
+	// upstreamServiceHost := "eu.auth0.com"
+
+	cluster := url.Hostname()
+	upstreamServiceHost := url.Hostname()
+	upstreamServicePort := uint32(443)
+	if "" != url.Port() {
+		port, err := strconv.ParseUint(url.Port(), 10, 32)
+		if err != nil {
+			err = fmt.Errorf("auth.NewFilterHTTPOAuth2: could not convert port=%q to int oauth2.token_endpoint=%q, %w", url.Port(), oauth2Options.TokenEndpoint, err)
+			args.Logger.Error(err, "oauth2.token_endpoint contains an invalid port, failed to parse url")
+			return nil, err
+		}
+		upstreamServicePort = uint32(port)
+	}
+
+	if !args.EnvoyConfiguration.ClusterExist(cluster) {
+		err := args.EnvoyConfiguration.AddClusterWithTLS(cluster, upstreamServiceHost, upstreamServicePort)
+		if err != nil {
+			return nil, fmt.Errorf("auth.NewFilterHTTPOAuth2: failed on `arguments.EnvoyConfiguration.AddClusterWithTLS`, err=%w", err)
+		}
+		// args.EnvoyConfiguration.AddCluster(cluster, upstreamServiceHost, upstreamServicePort)
+	} else {
+		fmt.Println("----------------")
+		fmt.Printf("auth.NewFilterHTTPOAuth2\n")
+		fmt.Println("")
+		fmt.Printf("cluster=%q exists\n", cluster)
+		fmt.Printf("upstreamServiceHost=%q\n", upstreamServiceHost)
+		fmt.Printf("upstreamServicePort=%v\n", upstreamServicePort)
+		fmt.Printf("cluster=%q\n", cluster)
+		// fmt.Printf("sni=%q\n", sni)
+		fmt.Println("----------------")
+	}
```

Comment:
I think this else block can be removed

Reply:
I'd rather keep this log line here, if that OK's with you, @kylehodgetts.

Reply:
Can't you put this into the logger and label it as debug so it is easy to turn on/off

Reply:
If it must stay then it should be part of a logger with the proper level. Error level if something went wrong

Reply:
It's a a debug level statement now.

It can't be an issue if two clusters exists, it was just a log line I added to help be figure out what was going on. Case in point, if you have two `paths`, like so:

```yaml
    paths:
      "/":
        get:
          description: Returns GET data.
          operationId: "/get"
          responses: {}
      "/uuid":
        get:
          description: Returns UUID4.
          operationId: "/uuid"
          responses: {}
```

then two OAuth2 filters are created, so on the second round the cluster already exists.

Reply:
Either way instead of using fmt.Println can't you use logger to print this out. Having debug println in the code doesn't make us look good.

Reply:
`args.Logger.V(1).Info` is a debug level log statement. Marking as resolved.